tests/unittests/test_backoff.py

```diff
@@ -0,0 +1,87 @@
+import unittest
+from unittest import mock
+from parameterized import parameterized
+from requests.exceptions import Timeout, ConnectionError
+from tap_quickbooks.client import Quickbooks4XXException, QuickbooksClient, Quickbooks5XXException,\
+    QuickbooksClient, QuickbooksBadRequestError, QuickbooksAuthenticationError, QuickbooksForbiddenError, QuickbooksNotFoundError, QuickbooksInternalServerError, QuickbooksServiceUnavailableError
+from test_exception_handling import get_response
+
+class TestBackoffOAuth2SessionInitialization(unittest.TestCase):
+    config = {
+        'refresh_token': 'token',
+        'client_id': 'id',
+        'client_secret': 'secret',
+        'user_agent': 'agent',
+        'realm_id': 'realm_id',
+        'sandbox': True
+    }
+
+    @parameterized.expand([
+        ['quickbooks_400_exception', [400, QuickbooksBadRequestError], [False, 3]],
+        ['quickbooks_401_exception', [401, QuickbooksAuthenticationError], [True, 3]],
+        ['quickbooks_402_exception', [402, Quickbooks4XXException], [False, 3]],
+        ['quickbooks_403_exception', [403, QuickbooksForbiddenError], [True, 3]],
+        ['quickbooks_404_exception', [404, QuickbooksNotFoundError], [False, 3]],
+        ['quickbooks_500_exception', [500, QuickbooksInternalServerError], [False, 3]],
+        ['quickbooks_502_exception', [502, Quickbooks5XXException], [False, 3]],
+        ['quickbooks_503_exception', [503, QuickbooksServiceUnavailableError], [False, 3]],
+        ['requests_connection_error', [None, ConnectionError], [False, 3]],
+        ['timeout_error', [None, Timeout], [False, 5]],
+    ])
+    @mock.patch('time.sleep')
+    @mock.patch('tap_quickbooks.client.QuickbooksClient._write_config')
+    @mock.patch('requests_oauthlib.OAuth2Session.request')
+    @mock.patch('tap_quickbooks.client.LOGGER.info')
```

Comment:
Is this the correct order for the decorators? If so, is there a way to make this more clear? test_backoff_for_OAuth2Session has arguments in an order that doesn't quite match the above given that the parameterized expand would be test_name, test_data, expected_data but then sleep through logger are in reversed order from the function args

Reply:
Yes, the order is correct.
Using parametrized expand, multiple test cases will be created during execution. Function args are given in the same order as in the @parametrized.expand decorator. Ref - [link](https://pypi.org/project/parameterized/#:~:text=.TestCase)%3A-,%40parameterized.expand,-(%5B%0A%20%20%20%20%20%20%20(%22negative)
For @mock.patch decorators, it will follow the bottom-up approach. So the first function arg will relate to the lowermost patch decorator.

Any suggestions to make it more clear?

Reply:
I don't have any ideas tbh because a comment feels clunky. I guess I just haven't seen parameterized used with decorators or used that many decorators before 🤷